lib/Controller/SettingsController.php

```diff
@@ -82,87 +90,113 @@ public function __construct($AppName, IRequest $request,
 		$this->accessTokenMapper = $accessTokenMapper;
 		$this->refreshTokenMapper = $refreshTokenMapper;
 		$this->userId = $UserId;
+		$this->l10n = $l10n;
 		$this->logger = $logger;
 		$this->urlGenerator = $urlGenerator;
 	}
 
 	/**
 	 * Adds a client.
 	 *
-	 * @return RedirectResponse Redirection to the settings page.
+	 * @return JSONResponse
 	 */
 	public function addClient() {
-		if (!isset($_POST['redirect_uri']) || !isset($_POST['name'])) {
-			return new RedirectResponse(
-				$this->urlGenerator->linkToRouteAbsolute(
-					'settings.SettingsPage.getAdmin',
-					['sectionid' => 'authentication']
-				) . '#oauth2');
+		$redirectUri = $this->request->getParam('redirect_uri', '');
+		$name = $this->request->getParam('name', '');
+		try {
+			if ($name === '') {
+				throw new \RuntimeException(
+					$this->l10n->t('Name must not be empty')
+				);
+			}
+			if ($redirectUri === '') {
+				throw new \RuntimeException(
+					$this->l10n->t('Redirect URI must not be empty')
+				);
+			}
+			if (!Utilities::isValidUrl($redirectUri)) {
+				throw new \RuntimeException(
+					$this->l10n->t('Redirect URI must be a valid URL')
+				);
+			}
+
+			try {
+				$this->clientMapper->findByName($name);
+				throw new \RuntimeException(
+					$this->l10n->t('Name %s already exists', [$name])
+				);
+			} catch (DoesNotExistException $e) {
+				// expected when the client name is not duplicated
+			}
+
+			$client = new Client();
+			$client->setIdentifier(Utilities::generateRandom());
+			$client->setSecret(Utilities::generateRandom());
+			$client->setRedirectUri(\trim($redirectUri));
+			$client->setName(\trim($name));
+
+			$allowSubdomains = $this->request->getParam('allow_subdomains', null) !== null;
+			$client->setAllowSubdomains($allowSubdomains);
+
+			$this->clientMapper->insert($client);
+			$this->logger->info('The client "' . $client->getName() . '" has been added.', ['app' => $this->appName]);
+
+			$template = new Template('oauth2', 'client.part', '');
+			$template->assign('client', $this->clientMapper->findByIdentifier($client->getIdentifier()));
+			$response = [
+				'status' => 'success',
+				'rowHtml' =>  $template->fetchPage(),
+				'data' => [] // OC.msg needs this
+			];
+		} catch (\RuntimeException $e) {
```

Comment:
yes, we should. If you prefer a bunch of nested `ifs` I surely would be glad to refactor the code into such a mess...
Otherwise I throw runtime exception if input argument does not meet any specific condition and catch it here to pass the error to the frontend.

Reply:
I'm not convinced about the goodies of this solution... I don't think it's so common, and the fact that I had problems to understand what you're trying to do here the first time, it lowers the score for me...
I mean, you normally throw an exception because you can't handle it.

What about something like:
```
if ($redirectUri === '') {
  return new JsonResponse([
    'status' => 'error',
    'errorMessage' => $this->l10n->t('Redirect URI must not be empty'),
    'data' => []
  ]);
}
if (!Utilities::isValidUrl($redirectUri)) {
  return new JsonResponse([
    'status' => 'error',
    'errorMessage' => $this->l10n->t('Redirect URI must be a valid URL'),
    'data' => []
  ]);
}
........
```

Reply:
@jvillafanez Such approach has the following disadvantages: more lines of code and multiple return points. But ok.